librz/bin/dwarf.c

```diff
@@ -2262,14 +2354,105 @@ RZ_API HtUP /*<offset, RzBinDwarfLocList *>*/ *rz_bin_dwarf_parse_loc(RzBinFile
 	return loc_table;
 }
 
-static void free_loc_table_entry(HtUPKv *kv) {
+static void ht_loc_list_free(HtUPKv *kv) {
 	if (kv) {
-		free_loc_table_list(kv->value);
+		loc_list_tree(kv->value);
 	}
 }
 
 RZ_API void rz_bin_dwarf_loc_free(HtUP /*<offset, RzBinDwarfLocList *>*/ *loc_table) {
-	rz_return_if_fail(loc_table);
-	loc_table->opt.freefn = free_loc_table_entry;
+	if (!loc_table) {
+		return;
+	}
+	loc_table->opt.freefn = ht_loc_list_free;
 	ht_up_free(loc_table);
 }
+
+RZ_API RZ_OWN RzBinDwarf *rz_bin_dwarf_parse(RZ_BORROW RZ_NONNULL RzBinFile *bf, RZ_BORROW RZ_NONNULL const RzBinDwarfParseOptions *opt) {
+	rz_return_val_if_fail(bf && opt, NULL);
+	RzBinDwarf *dw = RZ_NEW0(RzBinDwarf);
+	if (!dw) {
+		return NULL;
+	}
+
+	if (opt->flags & RZ_BIN_DWARF_PARSE_ABBREVS) {
+		RZ_LOG_DEBUG(".debug_abbrev\n");
+		dw->abbrevs = rz_bin_dwarf_abbrev_parse(bf);
+	}
+	if (opt->flags & RZ_BIN_DWARF_PARSE_INFO && dw->abbrevs) {
+		RZ_LOG_DEBUG(".debug_info\n");
+		dw->info = rz_bin_dwarf_info_parse(bf, dw->abbrevs);
+	}
+	if (opt->flags & RZ_BIN_DWARF_PARSE_LOC) {
+		RZ_LOG_DEBUG(".debug_loc\n");
+		dw->loc = rz_bin_dwarf_loc_parse(bf, opt->addr_size);
+	}
+	if (opt->flags & RZ_BIN_DWARF_PARSE_LINES && dw->info) {
+		RZ_LOG_DEBUG(".debug_line\n");
+		dw->lines = rz_bin_dwarf_parse_line(bf, dw->info, opt->line_mask);
+	}
+	if (opt->flags & RZ_BIN_DWARF_PARSE_ARANGES) {
+		RZ_LOG_DEBUG(".debug_aranges\n");
+		dw->aranges = rz_bin_dwarf_aranges_parse(bf);
+	}
+	return dw;
+}
+
+RZ_API void rz_bin_dwarf_free(RZ_OWN RzBinDwarf *dw) {
+	if (!dw) {
+		return;
+	}
+	rz_bin_dwarf_abbrev_free(dw->abbrevs);
+	rz_bin_dwarf_info_free(dw->info);
+	rz_bin_dwarf_loc_free(dw->loc);
+	rz_bin_dwarf_line_info_free(dw->lines);
+	rz_list_free(dw->aranges);
+	free(dw);
+}
+
+RZ_API RzBinDwarfAttr *rz_bin_dwarf_die_get_attr(const RzBinDwarfDie *die, enum DW_AT name) {
```

Comment:
Doxygen here and below

Reply:
Also ownership for functions that return pointers, and `RZ_NOTNULL`/`RZ_NULLABLE` for pointers like `die`